### ISSUE TITLE
SOAP message return carriage handling

### ISSUE THREAD

**Issue by bmaschio** [jolie/jolie#44]
@jolie/developers 
Hi all
While i was working on issue number #38 
I have noticed a problem while testing the SOAP call with soupUI with the following parameters

![image](https://cloud.githubusercontent.com/assets/12083716/7498085/9e3a7d38-f41e-11e4-8128-ca0821276f0c.png)

this is the error i get 
```xml
AVVERTENZA: [main_test_wsdl.ol] Received message TypeMismatch (input operation G
etCitiesByCountry): Invalid native type for node #Message.parameters: expected VOID, found java.lang.String
```
that is given by the fact that the good old soapUi sends carriage return characters 
Not even the 

```xml
 .dropRootValue = true
```
has desired effect because it is applied only on the root node 

I think this problem  could be solved by implementing  specification 2.11 End-of-Line Handling see link  
http://www.w3.org/TR/REC-xml/#sec-line-ends

I will have a go with it tomorrow

What do you think about it 

_b

**Comment by bmaschio** [comment]
@jolie/developers 

Maybe the issue a bit more simple then expected the original code was  

```xml  
               case Node.TEXT_NODE:
                    if (!(isRecRoot)) {
                        value.setValue(currNode.getNodeValue());
                    }
                break;
```
the new code

```xml 
                case Node.TEXT_NODE:
                    if (!(currNode.getNodeValue().trim().isEmpty())) {
                        value.setValue(currNode.getNodeValue());
                    }
                    break;
```
in this way it should work in the case of  a client like soapUi that send empty root with carriage return and also will be in line with jolie2wsdl rule that force void type for the root

```xml 
type GetCitiesByCountry:void{
	.CountryName?:string
}


type GetCitiesByCountryResponse:void{
	.GetCitiesByCountryResult?:string  /// parameters<= GetCitiesByCountryResponse.
}
/*
<wsdl:message name="GetCitiesByCountrySoapIn">
<wsdl:part name="parameters" element="tns:GetCitiesByCountry"/>
</wsdl:message>
<wsdl:message name="GetCitiesByCountrySoapOut">
<wsdl:part name="parameters" element="tns:GetCitiesByCountryResponse"/>
</wsdl:message>
*/

type GetCitiesByCountrySoapIn:void{
	.parameters: GetCitiesByCountry
}

type GetCitiesByCountrySoapOut:void{
	.parameters: GetCitiesByCountryResponse
}

```

now if the client sends a string where it is suppose to be void we still get 

![image](https://cloud.githubusercontent.com/assets/12083716/7502922/d1b88c56-f43f-11e4-93dc-49558b454056.png)

But then if there an empty string now it works

![image](https://cloud.githubusercontent.com/assets/12083716/7503005/48bf7cd8-f440-11e4-9e86-cf9198d9f85b.png)

I think we could also drop the parameter boolean isRecRoot

```xml 
private void xmlNodeToValue(Value value, Node node, boolean isRecRoot)

```

What do you think about it

thanks

**Comment by fmontesi** [comment]
Thanks Balint, good catch.

I don't know whether that's a good default behaviour though, it will break protocol transparency (we don't trim strings in other protocols).
It seems that Java has an official way of doing it:

http://docs.oracle.com/javase/7/docs/api/javax/xml/parsers/DocumentBuilderFactory.html#setIgnoringElementContentWhitespace(boolean)

The parser needs to be in "validation" mode for that, meaning that it should have the schema (or wsdl) of the service you're interacting with.

@klag @mwallnoefer : opinions?

**Comment by mwallnoefer** [comment]
@fmontesi indeed your approach seems to be cleaner. I would also drop the ".dropRootValue" parameter and introduce a ".ignoringElementContentWhitespace" one.

**Comment by bmaschio** [comment]
@fmontesi @mwallnoefer @klag

Ok i pick up and make the required modification

**Comment by bmaschio** [comment]
@fmontesi @mwallnoefer @klag 

shell we introduce something like 

```xml  
.protocol: soap{
 .validationMode: bool{
     .ignoringElementContentWhitespace:bool
 }
}
```

this should give us flexibility if we want to add other validation parameters  what you think?

**Comment by klag** [comment]
I really do not like the idea to have a parameter in soap protocol for stating to ignore line feeds. I think the standard way should be ignoring the TEXT NODES with carriage returns and line feed (http://www.w3.org/TR/REC-xml/#sec-line-ends) because this covers the 99% of real cases. Maybe, we could insert a parameter to enable the parsing of the TEXT NODES in case this is required.

The main idea is to parse the text node of an element only if:
- it is a final element whose type is a native type
- it is defined with a mixed content (http://www.w3schools.com/schema/schema_complex_mixed.asp)

How we could map a mixed content in jolie?

- wsdl2jolie --> mixed type = any
- jolie2wsdl --> any --> mixed content (only if there are subnodes)

The perfect way should be:

- wsdl2jolie: mixed type --> string | void  (where | is an or)
- jolie2wsdl: string | void or string --> mixed content  (only if there are subnodes). If the native value of the root is different from string | void or string, the jolie2wsdl should raise an error

**Comment by bmaschio** [comment]
@klag @fmontesi 
Maybe I have been not very clear on my approach. Yesterday @klag has correctly suggested to implement a standard  2.11. My  first idea was to act on  *xmlNodeToValue* method where in case of textNode check if carriage return code was present. 
Then I realized that SoapUi was also sending some spaces and I opted for ( brutal trim ) 
Rightly So @fmontesi  as suggested not to act on the protocol trimming it but using the Document Builder options 
I think @klag  idea of mixed type i **rather** cool because it also in my opinion will solve the issue of WSDL elements declared a nullable.
I think @klag  idea of mixed type i **rather** cool because it also in my opinion will solve the issue of WSDL elements declared a nullable.
I think @fmontesi  idea will solve the current issue with clients that send "WhiteSpaces" that has an impact on the usability via SOAP of Jolie services

**Comment by mwallnoefer** [comment]
@bmaschio so how to proceed here? Have I understood you correctly that you would like to add the parameters mentioned in one of your last comments? For my part they seem sound and we could drop ".dropRootValue".

**Comment by fmontesi** [comment]
I like @bmaschio 's proposal. It covers what we all discussed (for @klag 's comments, we simply need to have .ignoringElementContentWhitespace set to true by default).

I don't think we need to introduce .validationMode, as we already have .schema and .wsdl, both of which give type information.

Let's talk about what should happen when a schema and/or a wsdl are not provided.

Following http://www.w3.org/TR/REC-xml/#sec-line-ends I think that we should trim only the end of the string, not both left and right. However, I'm concerned by the fact that the spec says that the XML processor does that job already, and we're using one. It's unclear to me if maybe the spec is already implemented by the Java XML processor, and we're receiving one single #xA character as specified in the link.

But hey, I may be paranoid here. ;-)

We should document this properly, with a big box somewhere in the docs. Trimming strings may mess up with user documents written in XML. What are we doing already in XmlUtils and FileService when we read XML files? Do we have the same behaviour?

**Comment by mwallnoefer** [comment]
As far as I know we do nothing in respect to trimming.

**Comment by bmaschio** [comment]
@fmontesi @fmontesi  @fmontesi If all of you agree i will implement the @fmontesi 's suggested implementation

**Comment by mwallnoefer** [comment]
Okay, so please propose a patch for review :).

**Comment by fmontesi** [comment]
@bmaschio I think you meant @mwallnoefer @klag @fmontesi. :-)

@klag Do you see any problems incoming in existing software? Other comments?

**Comment by bmaschio** [comment]
@mwallnoefer @fmontesi @klag the ignoringElementContentWhitespace has not solved the problem of this issue i suppose the problem carriege return is not a whitespace therefore when passed to jolie operation it see it as a string
I propose to implement the following  logic if ignoringElementContentWhitespace == true then  normalize all carriege return with space then i set the DocumentBuilder with ignoringElementContentWhitespace = true
What do you think

**Comment by fmontesi** [comment]
Have you specified a .wsdl or .schema? It's needed for that parameter in document builder to have an effect, judging from the javadoc. It says that validation mode should be on.

**Comment by bmaschio** [comment]
@fmontesi  yes 

```xml  

inputPort TestInterfaceSoapInputPort {
  Location: SOAP_TEST_LOCATION
  Protocol: soap{
  	 .wsdl = "./TestWsdl.wsdl";
  	 .ignoringElementContentWhitespace = true;
         .validate = 1;
  	.debug = true;
       .charset = "UTF-8";
       .compression = false
  	  }
  Interfaces: JolieWeatherInterface 
}
```

Am I getting lost in something rather small?

**Comment by fmontesi** [comment]
No, the Jolie code is OK, but it seems from SOAPProtocol.java that we're
not setting the document builder to validation mode. Am I misunderstanding
the API? Could you have a look?

**Comment by mwallnoefer** [comment]
@bmaschio one note to the charset parameter: it is unsupported on webservices (JSON-RPC, XML-RPC, SOAP) - the (outgoing) encoding there will be always UTF-8, although the incoming messages could be encoded differently eg. in UTF-16 which is correctly decoded by our parser.

The parameter exists only in HTTP and SODEP with a default of UTF-8. On HTTP it sometimes makes sense to serve content in latin1 or UTF-16, but usually it should never be set or changed.

**Comment by bmaschio** [comment]
@mwallnoefer Thanks
@fmontesi Ok I found out the issue the .validate parameter must be a child of schema node but it does not change the validation of the DocumentBuilder
I added now child validate  the on the soap protocol parameters.
Please notice that the error is reported in Italian have you got any idea why?

Avvertenza: la convalida è stata attivata, ma org.xml.sax.ErrorHandler non è sta
to impostato, il che potrebbe essere un errore. Il parser utilizzerà un ErrorHan
dler predefinito per visualizzare i primi 0 errori. Richiamare il metodo 'setErr
orHandler' per correggere questo problema.
Error: URI=null Line=1: Documento non valido: nessuna grammatica trovata.
Error: URI=null Line=1: L'elemento radice "soapenv:Envelope" del documento deve
corrispondere alla radice DOCTYPE "null".
mag 13, 2015 9:04:21 AM jolie.Interpreter logWarning
AVVERTENZA: [main_test_wsdl.ol] Received message TypeMismatch (input operation G
etCitiesByCountry): Invalid native type for node #Message: expected VOID, found
java.lang.String
mag 13, 2015 9:04:21 AM jolie.Interpreter logInfo

**Comment by fmontesi** [comment]
Sorry @bmaschio, I replied a while ago and then closed the window instead of pressing "Comment" apparently.. :-\

It seems that we're not setting the schema in the documentbuilder used in SOAPProtocol. The following code is commented out:

                /*
                 * Schema messageSchema = getRecvMessageValidationSchema(); if (
                 * messageSchema != null ) {
                 * factory.setIgnoringElementContentWhitespace( true );
                 * factory.setSchema( messageSchema ); }
                 */

What gives? Anybody @jolie/developers ? I'll try git blame as soon as I can.

**Comment by fmontesi** [comment]
Mmmh and it looks that commit 7e188b15b19f54631caa3a8e732ca50372b6ef00 broke indentation... I'll get around to that too.

**Comment by bmaschio** [comment]
@fmontesi 
The internal method getRecvMessageValidationSchema is also missing. I will pick it up next few day been very busy with the new job.

**Comment by klag** [comment]
@bmaschio could you summarize me if this bug is solved or not? I tried to rebuild from the previous chat but I am a little bit confused. The problem you post at the beginning is still present?

**Comment by bmaschio** [comment]
@klag No I have not solved it it is still present

**Comment by bmaschio** [comment]
any news?

**Comment by fmontesi** [comment]
Not that  I know of. @jolie/developers anyone working on this?

**Comment by fmontesi** [comment]
If anybody is interested and needs some guidance, please feel free to ask!
